core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java

```diff
@@ -1110,6 +1110,21 @@ private RecordCursorFactory generateSelectChoose(QueryModel model, SqlExecutionC
     private RecordCursorFactory generateSelectDistinct(QueryModel model, SqlExecutionContext executionContext) throws SqlException {
         final RecordCursorFactory factory = generateSubQuery(model, executionContext);
         try {
+            if (model.getColumns().size() == 1 && model.getNestedModel() != null && model.getNestedModel().getNestedModel() != null && model.getNestedModel().getNestedModel().getTableName() != null) {
+                ExpressionNode tableName = model.getNestedModel().getNestedModel().getTableName();
+                TableReader reader = engine.getReader(executionContext.getCairoSecurityContext(), tableName.token);
```

Comment:
reader is a shared resource - SQL code generator must not assign it to the factory, but close it instead in try() block.